### ISSUE TITLE
[RFC] Methods to change theme and request zone

### ISSUE THREAD

**Issue by ToBe998** [bolt/bolt#5588]
Currently, Bolt uses one theme for all pages rendered by the frontend. In many cases, this is not sufficient. A page needs to be displayed in various variants, requiring different template sets.

Some examples:
- **Google Accelerated pages** display a page with a special set of html tags.
- **Newsletters** Newsletters have a very different html layout as your normal pages.
- **Special "skins"** Some websites allow the visitor to choose different skins for their pages.

It may be possible to use different html just by adding the login inside your one theme's twig, but this makes your theme overly complex **and wont solve differences in injected assets**.
So, an important additional requirement would be to modify injected assets. A Google Amp page does not allow bolt's assets or css/js from extensions to be included. Newsletters need absolute paths, and so on.

**What is needed**
You can already create your own controller, create a subrequest and relay that one to the normal frontend controller. So, in theory, you dont need to duplicate the frontend controller for each and every one of your variants.
Sadly, if you set the theme before relaying, it will be reset again.
Also, if you set a different zone in your subrequest, the asset queue's listeners will react to the main request and not the subrequest, still adding it's assets.

**TL;DR**
Please provide a clean way of changing the theme and the request's zone manually and provide a way od relaying the request to the frontend controller while respecting the changed values.

**Comment by GawainLynch** [comment]
ping @CarsonF

**Comment by ToBe998** [comment]
Note: In Bolt 3.0.6 you could at least hack the theme folder like this:
```$app['twig.loader.filesystem']->prependPath($path);
```
Now Bolt uses it’s own filesystem and that doesnt seem to work anymore. My best try so far is this, but it just doesnt correctly update the mounted paths…
```$filesystem = $app['filesystem'];
$filesystem->mountFilesystem("theme", new Filesystem(new Local($app['resources']->getPath('themebase') . '/' . $newtheme);
```

The asset rendering seems to be hackable like this:
```
            $request = &$app["request"];
            Zone::set($request, $config["themes"][$theme]["zone"]);
```
Also not nice but at least possible.

**Comment by ToBe998** [comment]
Update: This evil hack (i feel very dirty now) did the job for now. Having a clean solution would be great anyways.

```
$app["filesystem"]->mountFilesystem("theme", new Filesystem(new Local( $app['resources']->getPath('themebase')."/".$folder )));
$app["twig.loader.bolt_filesystem"]->prependPath( "theme://" );
```

**Comment by CarsonF** [comment]
@ToBe998 What does that hack do for you?

**Comment by ToBe998** [comment]
That hack changes the theme depending on the url.

 * The them is changed to a different one
 * Assets from the normal frontend are blocked if needed (Some themes like amp break if we load them)
* The request is relayed to frontend as usual

For example this url `/customtheme/contenttype/slug` sets the following stuff:
 * `general/theme` is set to `customtheme` (from a config of course, not directly from url)
 * the theme path is loaded with $app["filesystem]->mountFilesystem("theme", new Filesystem( new Local("my custom theme path")))`
 * The mount points inside bolts theme filesystem is updated with `$app["twig.loader.bolt_filesystem"]->prependPath( "theme://" )
 * The zone of the request is set to the configured zone with `Zone::set($request, $config["themes"][$theme]["zone"])`
 * the request is relayed to the normal frontend with `$app->handle(...)`

**Comment by ToBe998** [comment]
Just found out twig caches compiled templates based on path and filename. So if you add a second version of the same template, you will run into conflicts with cached versions.

So in addition to above points, i also had to change the cache folder for the switched theme.
`$app["twig"]->setCache( $app["twig"]->getCache()."/myswitchedtheme" )`

**Comment by bobdenotter** [comment]
Because of "Spring cleaning", we're going to close this issue. Unless somebody steps up to help out with implementing it, it's not expected that the core team will have the opportunity to work on this soon. Keeping it open might set false expectations for people stumbling upon this issue.

This is not because we don't see this feature as potentially worthwhile, it is more to do with an overload of potential new features that could be developed combined with non-limitless amounts of available time and resources. 

If you disagree, feel free to re-open for discussion.
